rust/src/quic/frames.rs

```diff
@@ -115,13 +123,210 @@ impl fmt::Display for StreamTag {
+                            value.extend_from_slice(alpn);
+                            values.push(value);
+                        }
+                    }
+                    _ => {}
+                }
+                extv.push(QuicTlsExtension { etype, values })
+            }
+            if client {
+                match write!(ja3, ",") {
+                    _ => {}
+                }
+                dash = false;
+                for e in &exts {
+                    match e {
+                        TlsExtension::EllipticCurves(x) => {
+                            for ec in x {
+                                if dash {
+                                    match write!(ja3, "-") {
+                                        _ => {}
+                                    }
+                                } else {
+                                    dash = true;
+                                }
+                                match write!(ja3, "{}", ec.0) {
+                                    _ => {}
+                                }
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                match write!(ja3, ",") {
+                    _ => {}
+                }
+                dash = false;
+                for e in &exts {
+                    match e {
+                        TlsExtension::EcPointFormats(x) => {
+                            for ec in *x {
+                                if dash {
+                                    match write!(ja3, "-") {
+                                        _ => {}
+                                    }
+                                } else {
+                                    dash = true;
+                                }
+                                match write!(ja3, "{}", ec) {
+                                    _ => {}
+                                }
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+    }
+    return extv;
+}
+
+fn parse_crypto_frame(input: &[u8]) -> IResult<&[u8], Frame, QuicError> {
+    let (rest, offset) = quic_var_uint(input)?;
+    let (rest, length) = quic_var_uint(rest)?;
+    let (rest, _) = take(offset as usize)(rest)?;
+    let (_, data) = take(length as usize)(rest)?;
+
+    if let Ok((rest, msg)) = parse_tls_message_handshake(data) {
+        if let Handshake(hs) = msg {
+            match hs {
+                ClientHello(ch) => {
+                    let mut ja3 = String::with_capacity(256);
+                    match write!(&mut ja3, "{},", u16::from(ch.version)) {
+                        _ => {}
+                    }
+                    let mut dash = false;
+                    for c in &ch.ciphers {
+                        if dash {
+                            match write!(&mut ja3, "-") {
```

Comment:
these kinds of statements are taking 3 lines each for no good reason. Can these be made more compact?

Reply:
```
let _ = write...
```
(noted above already)

Reply:
`push_str` seems nice